feat: add --write-ask option to prompt for each change

Bnyro opened the pull request:
closes #397

epage commented on crates/typos-cli/src/bin/typos-cli/args.rs:
```diff
@@ -67,6 +67,10 @@ pub(crate) struct Args {
 
+    /// Prompt for each suggested correction whether to write the fix
+    #[arg(long, short = 'a', group = "mode", help_heading = "Mode")]
+    pub(crate) write_ask: bool,
```
Generally this would be `-i` `--interactive`

epage commented on crates/typos-cli/src/file.rs:
```diff
@@ -127,12 +129,85 @@ impl FileChecker for FixTypos {
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct AskFixTypos;
```
Again, `Interactive`

epage commented on crates/typos-cli/src/file.rs:
```diff
@@ -127,12 +129,85 @@ impl FileChecker for FixTypos {
-                    let new_name =
-                        String::from_utf8(new_name).expect("corrections are valid utf-8");
-                    let new_path = path.with_file_name(new_name);
-                    std::fs::rename(path, new_path)?;
+                    fix_file_name(path, &file_name.to_owned(), fixes)?;
+                }
+            }
+        }
+
+        Ok(())
+    }
```
Please split refactors out into their own commits before the feature work

epage commented on crates/typos-cli/src/file.rs:
```diff
@@ -127,12 +129,85 @@ impl FileChecker for FixTypos {
+            if let Some(file_name) = path.file_name().and_then(|s| s.to_str()) {
+                let mut fixes = Vec::new();
+
+                for typo in check_str(file_name, policy) {
+                    let msg = report::Typo {
+                        context: Some(report::PathContext { path }.into()),
+                        buffer: std::borrow::Cow::Borrowed(file_name.as_bytes()),
+                        byte_offset: typo.byte_offset,
+                        typo: typo.typo.as_ref(),
+                        corrections: typo.corrections.clone(),
+                    };
+                    reporter.report(msg.into())?;
+
+                    if let Some(correction_index) = select_fix(&typo) {
```
We're only supposed to report things that aren't fixed

Bnyro replied:
"not fixed" means typos that only have one possible correction? Shouldn't we prompt whether the user wants to apply that single possible fix?

Or do you mean running `reporter.report(msg.into())?`? That part however displays the suggested fix to the user, doesn't it?

epage replied:
reporting is meant for use with `--format`.  It seems you are trying to take advantage of the output formatting for outputting the part of the prompt.  Technically, this will break down for json output but people probably shouldn't use this for progammatic output.

We should at least have a `// HACK:` comment explaining the abuse of reporting for the sake of prompting the user and that technically we should only report when not correcting.

Bnyro replied:
> It seems you are trying to take advantage of the output formatting for outputting the part of the prompt. 
>
Oh well, I see, I didn't notice it's possible to specify other output format than the `long` format.

epage commented on crates/typos-cli/src/file.rs:
```diff
@@ -127,12 +129,85 @@ impl FileChecker for FixTypos {
+                        fixes.push((typo, correction_index));
+                    }
+
+                    println!("\n");
```
Why this newline?

Bnyro replied:
To clearly have a visual separation between the prompt/selection about the current fix and the previous fix. Can be removed though if you fix it's unnecessary.

epage commented on crates/typos-cli/src/file.rs:
```diff
@@ -650,6 +725,40 @@ fn is_fixable(typo: &typos::Typo<'_>) -> bool {
     extract_fix(typo).is_some()
 }
 
+fn fix_buffer_with_correction_index<'a>(
+    mut buffer: Vec<u8>,
+    typos: impl Iterator<Item = (typos::Typo<'a>, usize)>,
+) -> Vec<u8> {
+    let mut offset = 0isize;
+    for (typo, correction_index) in typos {
+        let fix = match &typo.corrections {
+            typos::Status::Corrections(c) => Some(c[correction_index].as_ref()),
+            _ => None,
+        }
+        .expect("Caller provided invalid fix index");
+        let start = ((typo.byte_offset as isize) + offset) as usize;
+        let end = start + typo.typo.len();
+
+        buffer.splice(start..end, fix.as_bytes().iter().copied());
+
+        offset += (fix.len() as isize) - (typo.typo.len() as isize);
+    }
+    buffer
+}
+
+fn fix_file_name_with_correction_index<'a>(
+    path: &std::path::Path,
+    file_name: &'a str,
+    fixes: Vec<(typos::Typo<'a>, usize)>,
+) -> Result<(), std::io::Error> {
+    let file_name = file_name.to_owned().into_bytes();
+    let new_name = fix_buffer_with_correction_index(file_name, fixes.into_iter());
+    let new_name = String::from_utf8(new_name).expect("corrections are valid utf-8");
+    let new_path = path.with_file_name(new_name);
+    std::fs::rename(path, new_path)?;
+    Ok(())
+}
```
Feel free to do so.  Please generalize the function in a refactor commit before the feature work

epage commented on crates/typos-cli/src/file.rs:
```diff
@@ -664,6 +773,56 @@ fn fix_buffer(mut buffer: Vec<u8>, typos: impl Iterator<Item = typos::Typo<'stat
+}
+
+fn select_fix(typo: &typos::Typo<'_>) -> Option<usize> {
+    if is_fixable(&typo) {
```
Instead of using `is_fixable`, extract the corrections and condition this on the length.

epage commented on crates/typos-cli/src/file.rs:
```diff
@@ -664,6 +773,56 @@ fn fix_buffer(mut buffer: Vec<u8>, typos: impl Iterator<Item = typos::Typo<'stat
+            typos::Status::Corrections(c) => c,
+            _ => return None,
+        }
+        .clone();
+        items.insert(0, std::borrow::Cow::from("None (skip)"));
+
+        let selection = Select::new()
+            .with_prompt("Please choose one of the following suggestions")
+            .items(&items)
+            .default(0)
+            .interact()
+            .unwrap();
+
+        if selection == 0 {
+            return None;
+        }
+
+        return Some(selection - 1);
```
```suggestion
        .clone();

        items.insert(0, std::borrow::Cow::from("None (skip)"));
        let selection = Select::new()
            .with_prompt("Please choose one of the following suggestions")
            .items(&items)
            .default(0)
            .interact()
            .unwrap();
        if selection == 0 {
            return None;
        }
        return Some(selection - 1);
```

By grouping using blank lines, it helps clarify that these are all coupled together.  For example, I originally missed `items.insert` and was assuming `selection - 1` was a quirk of that API.

epage commented on crates/typos-cli/src/file.rs:
```diff
@@ -664,6 +773,56 @@ fn fix_buffer(mut buffer: Vec<u8>, typos: impl Iterator<Item = typos::Typo<'stat
+    Ok(())
+}
+
+fn select_fix(typo: &typos::Typo<'_>) -> Option<usize> {
```
When I run this locally, I get weird output on the right item

![image](https://github.com/user-attachments/assets/61de492d-fb84-4ceb-be6f-972253078cd8)

epage replied:
Something still isn't right
![image](https://github.com/user-attachments/assets/3feeb35c-3d4a-4b58-a7b6-410931794172)

And if I ctrl-c,. my cursor no longer shows up

Bnyro replied:
I can reproduce the issue with the cursor (but thought something in my setup is broken, apparently it's not my setup), I've really no idea what causes it. I've tried the `inquire` library as well because I thought something is wrong with the `dialoguer` library but that one worked even worse...

Bnyro replied:
Oh well, apparently we have to handle Ctrl+C issues ourselves as per https://github.com/console-rs/dialoguer/issues/294.

Bnyro replied:
It seems like the available cli libraries for prompts are not really that mature yet.

Bnyro commented:
Thanks for your review so far, I'm way more comfortable with the code now than before the review.

epage commented on crates/typos-cli/src/file.rs:
```diff
@@ -650,10 +642,24 @@ fn is_fixable(typo: &typos::Typo<'_>) -> bool {
-fn fix_buffer(mut buffer: Vec<u8>, typos: impl Iterator<Item = typos::Typo<'static>>) -> Vec<u8> {
+fn fix_buffer(
+    mut buffer: Vec<u8>,
+    typos: Vec<typos::Typo<'_>>,
+    correction_indices: Option<Vec<usize>>,
```
I feel like instead of parallel vec's, we should just have one iterator with the typo and index in a tuple.  For the existing cases, we could pass in `0`.

epage commented on crates/typos-cli/src/file.rs:
```diff
@@ -664,6 +670,19 @@ fn fix_buffer(mut buffer: Vec<u8>, typos: impl Iterator<Item = typos::Typo<'stat
     buffer
 }
 
+fn fix_file_name<'a>(
```
Since you are already going back and making a change, could you split the extraction of `fix_file_name` out into its own commit before the adding of indices?  Each commit should do only one thing.  This is a big help for reviewers as I don't have to try to untangle the root reason for each change

epage commented on crates/typos-cli/src/file.rs:
```diff
@@ -683,6 +764,42 @@ fn fix_file_name<'a>(
+            return Some(0);
+        }
+    } else {
+        let mut items = corrections.clone();
```
nit: if we've cloned the corrections on 772, why do we need this clone?  We could just do `let mut items = corrections;`

Bnyro commented:
I've added a custom `ctrlc` handler now to handle manual program exits and added proper error handling to the `dialoguer` confirm and selection prompts.

I've no idea why the position of the prompt is weird sometimes.

epage commented on crates/typos-cli/src/bin/typos-cli/main.rs:
```diff
@@ -32,6 +32,13 @@ fn run() -> proc_exit::ExitResult {
+        let _ = console::Term::stdout().show_cursor();
+        std::process::exit(0);
+    })
+    .expect("ctrlc::set_handler");
```
an "expect" should say what assumption it is operating under for why its ok.

epage commented on crates/typos-cli/src/bin/typos-cli/main.rs:
```diff
@@ -32,6 +32,13 @@ fn run() -> proc_exit::ExitResult {
 
     init_logging(args.verbose.log_level());
 
+    // https://github.com/console-rs/dialoguer/issues/294
```
It would help to put in a bit more detail, like
```rust
// HACK: Ensure the terminal gets reset to a good state if the user hits ctrl-c during a prompt
// see https://github.com/console-rs/dialoguer/issues/294 for more details
```

epage commented on crates/typos-cli/src/file.rs:
```diff
@@ -664,6 +745,53 @@ fn fix_buffer(mut buffer: Vec<u8>, typos: impl Iterator<Item = typos::Typo<'stat
+    };
+
+    if corrections.len() == 1 {
+        let confirmation = Confirm::new()
+            .with_prompt("Do you want to apply the fix suggested above?")
+            .default(true)
+            .show_default(true)
+            .interact();
+        return confirmation.map(|_| 0).ok();
```
imo use of `?` would make this easier to read

```suggestion
        let confirmation = Confirm::new()
            .with_prompt("Do you want to apply the fix suggested above?")
            .default(true)
            .show_default(true)
            .interact()
            .ok()?;
        0
```

epage commented on crates/typos-cli/src/file.rs:
```diff
@@ -664,6 +745,53 @@ fn fix_buffer(mut buffer: Vec<u8>, typos: impl Iterator<Item = typos::Typo<'stat
+            .with_prompt("Please choose one of the following suggestions")
+            .items(&items)
+            .default(0)
+            .interact();
+        return match selection {
+            Ok(selection) => {
+                if selection != 0 {
+                    Some(selection - 1)
+                } else {
+                    None
+                }
+            }
+            Err(_) => None,
+        };
```
Similar
```suggestion
            .interact()
            .ok();
        let index = selection.checked_sub(1)?;
        index
```

epage commented on crates/typos-cli/src/bin/typos-cli/main.rs:
```diff
@@ -32,6 +32,13 @@ fn run() -> proc_exit::ExitResult {
 
+    // https://github.com/console-rs/dialoguer/issues/294
+    ctrlc::set_handler(move || {
+        let _ = console::Term::stdout().show_cursor();
```
I now see the cursor but I can't see anything I type

Bnyro replied:
That's strange, for me everything behaves normal after that change.

epage commented on crates/typos-cli/src/bin/typos-cli/main.rs:
```diff
@@ -32,6 +32,13 @@ fn run() -> proc_exit::ExitResult {
+    // https://github.com/console-rs/dialoguer/issues/294
+    ctrlc::set_handler(move || {
+        let _ = console::Term::stdout().show_cursor();
+        std::process::exit(0);
```
Is `exit(0)` the correct way of exiting on ctrl-c to mirror behavior before this change?

Bnyro replied:
You're right, the previous exit code for `ctrl-c` was 130, and according to my quick research 130 is preferred over 0 when `ctrl-c` killed the process via SIGTERM.

epage commented:
> I've no idea why the position of the prompt is weird sometimes.

For me, its always weird for the first prompt.  This would be a blocker for merging

Bnyro commented:
> For me, its always weird for the first prompt. This would be a blocker for merging
>
I see, but unfortunately I don't see any way to fix this except for trying other libraries?

epage commented:
How about we switch to showing prompts that require hitting `<enter>` like `git add -p` does?

Bnyro commented:
> How about we switch to showing prompts that require hitting `<enter>` like `git add -p` does?
>
That could fix the issue for the `confirm` prompt. If I run `typos --interactive` on this repository (crate-ci/typos), the first prompt is a selection prompt which is getting skipped immediately (due to the bug mentioned), although the `select` prompts should in theory require pressing `<enter>` as well. That's quite odd.

If I do however run this on different source code folders, for some everything works normally, and for some it behaves as described above, which doesn't make a lot of sense to me.

The only two considerable crates I found are the `dialoguer` crate (in use right now) and the `inquire` crate, which wouldn't even let me use the `Confirm` prompt properly, I wasn't able to actually enter yes or no when I tested it.
